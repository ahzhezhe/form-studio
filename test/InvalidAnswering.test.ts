import { Form } from '../src/Form';
import { validators, getConfigs } from '.';

describe('Invalid Answering', () => {

  const form = new Form(getConfigs(), validators);

  test('setAnswer undefined answer', () => {
    form.setAnswer('subGroup1Question1', undefined);

    expect(form.getRenderInstructions()).toMatchSnapshot();
    expect(form.getErrors()).toMatchSnapshot();
    expect(form.getCurrentAnswers()).toMatchSnapshot();
    expect(form.getValidatedAnswers()).toMatchSnapshot();
  });

  test('setChoice undefined answer', () => {
    form.setChoice('subGroup1Question2', undefined);

    expect(form.getRenderInstructions()).toMatchSnapshot();
    expect(form.getErrors()).toMatchSnapshot();
    expect(form.getCurrentAnswers()).toMatchSnapshot();
    expect(form.getValidatedAnswers()).toMatchSnapshot();
  });

  test('setChoices empty answer', () => {
    form.setChoices('subGroup1Question3', []);

    expect(form.getRenderInstructions()).toMatchSnapshot();
    expect(form.getErrors()).toMatchSnapshot();
    expect(form.getCurrentAnswers()).toMatchSnapshot();
    expect(form.getValidatedAnswers()).toMatchSnapshot();
  });

  test('setAnswer invalid question', () => {
    expect(() => form.setAnswer('invalid', 'subGroup1Question1Answer')).toThrow();
  });

  test('setChoice invalid question', () => {
    expect(() => form.setChoice('invalid', 'subGroup1Question2Choice1')).toThrow();
  });

  test('setChoices invalid question', () => {
    expect(() => form.setChoices('invalid', ['subGroup1Question3Choice1', 'subGroup1Question3Choice2'])).toThrow();
  });

  test('setChoice invalid choice', () => {
    form.setChoice('subGroup1Question2', 'invalid');

    expect(form.getRenderInstructions()).toMatchSnapshot();
    expect(form.getErrors()).toMatchSnapshot();
    expect(form.getCurrentAnswers()).toMatchSnapshot();
    expect(form.getValidatedAnswers()).toMatchSnapshot();
  });

  test('setChoices invalid choice', () => {
    form.setChoices('subGroup1Question3', ['invalid']);

    expect(form.getRenderInstructions()).toMatchSnapshot();
    expect(form.getErrors()).toMatchSnapshot();
    expect(form.getCurrentAnswers()).toMatchSnapshot();
    expect(form.getValidatedAnswers()).toMatchSnapshot();
  });

  test('selectChoice invalid choice', () => {
    expect(() => form.selectChoice('invalid', true)).toThrow();
  });

  test('setAnswer invalid question type', () => {
    expect(() => form.setAnswer('subGroup1Question2', 'subGroup1Question2Value')).toThrow();
  });

  test('setChoice invalid question type', () => {
    expect(() => form.setChoice('subGroup1Question1', 'subGroup1Question1Choice1')).toThrow();
  });

  test('setChoices invalid question type', () => {
    expect(() => form.setChoices('subGroup1Question1', ['subGroup1Question1Choice1', 'subGroup1Question1Choice2'])).toThrow();
  });

  test('clearGroup invalid group', () => {
    expect(() => form.clearGroup('invalid')).toThrow();
  });

  test('clearAnswer invalid question', () => {
    expect(() => form.clearAnswer('invalid')).toThrow();
  });

  test('resetGroup invalid group', () => {
    expect(() => form.resetGroup('invalid')).toThrow();
  });

  test('resetAnswer invalid question', () => {
    expect(() => form.resetAnswer('invalid')).toThrow();
  });

});
