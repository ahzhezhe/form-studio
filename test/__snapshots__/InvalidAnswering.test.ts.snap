// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invalid Answering setAnswer undefined answer 1`] = `
Array [
  Object {
    "custom": undefined,
    "disabled": false,
    "groups": Array [
      Object {
        "custom": undefined,
        "disabled": false,
        "groups": Array [],
        "id": "subGroup1",
        "questions": Array [
          Object {
            "choices": Array [],
            "currentAnswer": undefined,
            "custom": Object {
              "title": "subGroup1Questions",
            },
            "disabled": false,
            "error": "This field is required.",
            "id": "subGroup1Question1",
            "type": "any",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice1",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice1",
                "value": "subGroup1Question2Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice2",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice2",
                "value": "subGroup1Question2Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice3",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice3",
                "value": "subGroup1Question2Choice3",
              },
            ],
            "currentAnswer": undefined,
            "custom": Object {
              "title": "subGroup1Question2",
            },
            "disabled": false,
            "error": "Please select an option.",
            "id": "subGroup1Question2",
            "type": "choice",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice1",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice1",
                "value": "subGroup1Question3Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice2",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice2",
                "value": "subGroup1Question3Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice3",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice3",
                "value": "subGroup1Question3Choice3",
              },
            ],
            "currentAnswer": Array [],
            "custom": Object {
              "title": "subGroup1Question3",
            },
            "disabled": true,
            "error": "Please select at least 1 option.",
            "id": "subGroup1Question3",
            "type": "choices",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice1",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice1",
                "value": "subGroup1Question4Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice2",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice2",
                "value": "subGroup1Question4Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice3",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice3",
                "value": "subGroup1Question4Choice3",
              },
            ],
            "currentAnswer": Array [],
            "custom": Object {
              "title": "subGroup1Question3",
            },
            "disabled": false,
            "error": "Please select at least 1 option.",
            "id": "subGroup1Question4",
            "type": "choices",
            "validatedAnswer": undefined,
            "validating": false,
          },
        ],
      },
    ],
    "id": "group1",
    "questions": Array [
      Object {
        "choices": Array [],
        "currentAnswer": undefined,
        "custom": Object {
          "title": "group1Questions",
        },
        "disabled": false,
        "error": "This field is required.",
        "id": "group1Question1",
        "type": "any",
        "validatedAnswer": undefined,
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "custom": Object {
              "title": "group1Question2Choice1",
            },
            "disabled": false,
            "id": "group1Question2Choice1",
            "value": "group1Question2Choice1",
          },
          Object {
            "custom": Object {
              "title": "group1Question2Choice2",
            },
            "disabled": false,
            "id": "group1Question2Choice2",
            "value": "group1Question2Choice2",
          },
          Object {
            "custom": Object {
              "title": "group1Question2Choice3",
            },
            "disabled": false,
            "id": "group1Question2Choice3",
            "value": "group1Question2Choice3",
          },
        ],
        "currentAnswer": undefined,
        "custom": Object {
          "title": "group1Question2",
        },
        "disabled": false,
        "error": "Please select an option.",
        "id": "group1Question2",
        "type": "choice",
        "validatedAnswer": undefined,
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "custom": Object {
              "title": "group1Question3Choice1",
            },
            "disabled": false,
            "id": "group1Question3Choice1",
            "value": "group1Question3Choice1",
          },
          Object {
            "custom": Object {
              "title": "group1Question3Choice2",
            },
            "disabled": false,
            "id": "group1Question3Choice2",
            "value": "group1Question3Choice2",
          },
          Object {
            "custom": Object {
              "title": "group1Question3Choice3",
            },
            "disabled": false,
            "id": "group1Question3Choice3",
            "value": "group1Question3Choice3",
          },
        ],
        "currentAnswer": Array [],
        "custom": Object {
          "title": "group1Question3",
        },
        "disabled": false,
        "error": undefined,
        "id": "group1Question3",
        "type": "choices",
        "validatedAnswer": Array [],
        "validating": false,
      },
    ],
  },
]
`;

exports[`Invalid Answering setAnswer undefined answer 2`] = `
Object {
  "group1Question1": "This field is required.",
  "group1Question2": "Please select an option.",
  "subGroup1Question1": "This field is required.",
  "subGroup1Question2": "Please select an option.",
  "subGroup1Question4": "Please select at least 1 option.",
}
`;

exports[`Invalid Answering setAnswer undefined answer 3`] = `
Object {
  "group1Question3": Array [],
  "subGroup1Question3": Array [],
  "subGroup1Question4": Array [],
}
`;

exports[`Invalid Answering setAnswer undefined answer 4`] = `
Object {
  "group1Question3": Array [],
}
`;

exports[`Invalid Answering setChoice invalid choice 1`] = `
Array [
  Object {
    "custom": undefined,
    "disabled": false,
    "groups": Array [
      Object {
        "custom": undefined,
        "disabled": false,
        "groups": Array [],
        "id": "subGroup1",
        "questions": Array [
          Object {
            "choices": Array [],
            "currentAnswer": undefined,
            "custom": Object {
              "title": "subGroup1Questions",
            },
            "disabled": false,
            "error": "This field is required.",
            "id": "subGroup1Question1",
            "type": "any",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice1",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice1",
                "value": "subGroup1Question2Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice2",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice2",
                "value": "subGroup1Question2Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice3",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice3",
                "value": "subGroup1Question2Choice3",
              },
            ],
            "currentAnswer": undefined,
            "custom": Object {
              "title": "subGroup1Question2",
            },
            "disabled": false,
            "error": "Please select an option.",
            "id": "subGroup1Question2",
            "type": "choice",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice1",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice1",
                "value": "subGroup1Question3Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice2",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice2",
                "value": "subGroup1Question3Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice3",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice3",
                "value": "subGroup1Question3Choice3",
              },
            ],
            "currentAnswer": Array [],
            "custom": Object {
              "title": "subGroup1Question3",
            },
            "disabled": true,
            "error": "Please select at least 1 option.",
            "id": "subGroup1Question3",
            "type": "choices",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice1",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice1",
                "value": "subGroup1Question4Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice2",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice2",
                "value": "subGroup1Question4Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice3",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice3",
                "value": "subGroup1Question4Choice3",
              },
            ],
            "currentAnswer": Array [],
            "custom": Object {
              "title": "subGroup1Question3",
            },
            "disabled": false,
            "error": "Please select at least 1 option.",
            "id": "subGroup1Question4",
            "type": "choices",
            "validatedAnswer": undefined,
            "validating": false,
          },
        ],
      },
    ],
    "id": "group1",
    "questions": Array [
      Object {
        "choices": Array [],
        "currentAnswer": undefined,
        "custom": Object {
          "title": "group1Questions",
        },
        "disabled": false,
        "error": "This field is required.",
        "id": "group1Question1",
        "type": "any",
        "validatedAnswer": undefined,
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "custom": Object {
              "title": "group1Question2Choice1",
            },
            "disabled": false,
            "id": "group1Question2Choice1",
            "value": "group1Question2Choice1",
          },
          Object {
            "custom": Object {
              "title": "group1Question2Choice2",
            },
            "disabled": false,
            "id": "group1Question2Choice2",
            "value": "group1Question2Choice2",
          },
          Object {
            "custom": Object {
              "title": "group1Question2Choice3",
            },
            "disabled": false,
            "id": "group1Question2Choice3",
            "value": "group1Question2Choice3",
          },
        ],
        "currentAnswer": undefined,
        "custom": Object {
          "title": "group1Question2",
        },
        "disabled": false,
        "error": "Please select an option.",
        "id": "group1Question2",
        "type": "choice",
        "validatedAnswer": undefined,
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "custom": Object {
              "title": "group1Question3Choice1",
            },
            "disabled": false,
            "id": "group1Question3Choice1",
            "value": "group1Question3Choice1",
          },
          Object {
            "custom": Object {
              "title": "group1Question3Choice2",
            },
            "disabled": false,
            "id": "group1Question3Choice2",
            "value": "group1Question3Choice2",
          },
          Object {
            "custom": Object {
              "title": "group1Question3Choice3",
            },
            "disabled": false,
            "id": "group1Question3Choice3",
            "value": "group1Question3Choice3",
          },
        ],
        "currentAnswer": Array [],
        "custom": Object {
          "title": "group1Question3",
        },
        "disabled": false,
        "error": undefined,
        "id": "group1Question3",
        "type": "choices",
        "validatedAnswer": Array [],
        "validating": false,
      },
    ],
  },
]
`;

exports[`Invalid Answering setChoice invalid choice 2`] = `
Object {
  "group1Question1": "This field is required.",
  "group1Question2": "Please select an option.",
  "subGroup1Question1": "This field is required.",
  "subGroup1Question2": "Please select an option.",
  "subGroup1Question4": "Please select at least 1 option.",
}
`;

exports[`Invalid Answering setChoice invalid choice 3`] = `
Object {
  "group1Question3": Array [],
  "subGroup1Question3": Array [],
  "subGroup1Question4": Array [],
}
`;

exports[`Invalid Answering setChoice invalid choice 4`] = `
Object {
  "group1Question3": Array [],
}
`;

exports[`Invalid Answering setChoice undefined answer 1`] = `
Array [
  Object {
    "custom": undefined,
    "disabled": false,
    "groups": Array [
      Object {
        "custom": undefined,
        "disabled": false,
        "groups": Array [],
        "id": "subGroup1",
        "questions": Array [
          Object {
            "choices": Array [],
            "currentAnswer": undefined,
            "custom": Object {
              "title": "subGroup1Questions",
            },
            "disabled": false,
            "error": "This field is required.",
            "id": "subGroup1Question1",
            "type": "any",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice1",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice1",
                "value": "subGroup1Question2Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice2",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice2",
                "value": "subGroup1Question2Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice3",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice3",
                "value": "subGroup1Question2Choice3",
              },
            ],
            "currentAnswer": undefined,
            "custom": Object {
              "title": "subGroup1Question2",
            },
            "disabled": false,
            "error": "Please select an option.",
            "id": "subGroup1Question2",
            "type": "choice",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice1",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice1",
                "value": "subGroup1Question3Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice2",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice2",
                "value": "subGroup1Question3Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice3",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice3",
                "value": "subGroup1Question3Choice3",
              },
            ],
            "currentAnswer": Array [],
            "custom": Object {
              "title": "subGroup1Question3",
            },
            "disabled": true,
            "error": "Please select at least 1 option.",
            "id": "subGroup1Question3",
            "type": "choices",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice1",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice1",
                "value": "subGroup1Question4Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice2",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice2",
                "value": "subGroup1Question4Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice3",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice3",
                "value": "subGroup1Question4Choice3",
              },
            ],
            "currentAnswer": Array [],
            "custom": Object {
              "title": "subGroup1Question3",
            },
            "disabled": false,
            "error": "Please select at least 1 option.",
            "id": "subGroup1Question4",
            "type": "choices",
            "validatedAnswer": undefined,
            "validating": false,
          },
        ],
      },
    ],
    "id": "group1",
    "questions": Array [
      Object {
        "choices": Array [],
        "currentAnswer": undefined,
        "custom": Object {
          "title": "group1Questions",
        },
        "disabled": false,
        "error": "This field is required.",
        "id": "group1Question1",
        "type": "any",
        "validatedAnswer": undefined,
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "custom": Object {
              "title": "group1Question2Choice1",
            },
            "disabled": false,
            "id": "group1Question2Choice1",
            "value": "group1Question2Choice1",
          },
          Object {
            "custom": Object {
              "title": "group1Question2Choice2",
            },
            "disabled": false,
            "id": "group1Question2Choice2",
            "value": "group1Question2Choice2",
          },
          Object {
            "custom": Object {
              "title": "group1Question2Choice3",
            },
            "disabled": false,
            "id": "group1Question2Choice3",
            "value": "group1Question2Choice3",
          },
        ],
        "currentAnswer": undefined,
        "custom": Object {
          "title": "group1Question2",
        },
        "disabled": false,
        "error": "Please select an option.",
        "id": "group1Question2",
        "type": "choice",
        "validatedAnswer": undefined,
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "custom": Object {
              "title": "group1Question3Choice1",
            },
            "disabled": false,
            "id": "group1Question3Choice1",
            "value": "group1Question3Choice1",
          },
          Object {
            "custom": Object {
              "title": "group1Question3Choice2",
            },
            "disabled": false,
            "id": "group1Question3Choice2",
            "value": "group1Question3Choice2",
          },
          Object {
            "custom": Object {
              "title": "group1Question3Choice3",
            },
            "disabled": false,
            "id": "group1Question3Choice3",
            "value": "group1Question3Choice3",
          },
        ],
        "currentAnswer": Array [],
        "custom": Object {
          "title": "group1Question3",
        },
        "disabled": false,
        "error": undefined,
        "id": "group1Question3",
        "type": "choices",
        "validatedAnswer": Array [],
        "validating": false,
      },
    ],
  },
]
`;

exports[`Invalid Answering setChoice undefined answer 2`] = `
Object {
  "group1Question1": "This field is required.",
  "group1Question2": "Please select an option.",
  "subGroup1Question1": "This field is required.",
  "subGroup1Question2": "Please select an option.",
  "subGroup1Question4": "Please select at least 1 option.",
}
`;

exports[`Invalid Answering setChoice undefined answer 3`] = `
Object {
  "group1Question3": Array [],
  "subGroup1Question3": Array [],
  "subGroup1Question4": Array [],
}
`;

exports[`Invalid Answering setChoice undefined answer 4`] = `
Object {
  "group1Question3": Array [],
}
`;

exports[`Invalid Answering setChoices empty answer 1`] = `
Array [
  Object {
    "custom": undefined,
    "disabled": false,
    "groups": Array [
      Object {
        "custom": undefined,
        "disabled": false,
        "groups": Array [],
        "id": "subGroup1",
        "questions": Array [
          Object {
            "choices": Array [],
            "currentAnswer": undefined,
            "custom": Object {
              "title": "subGroup1Questions",
            },
            "disabled": false,
            "error": "This field is required.",
            "id": "subGroup1Question1",
            "type": "any",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice1",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice1",
                "value": "subGroup1Question2Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice2",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice2",
                "value": "subGroup1Question2Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice3",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice3",
                "value": "subGroup1Question2Choice3",
              },
            ],
            "currentAnswer": undefined,
            "custom": Object {
              "title": "subGroup1Question2",
            },
            "disabled": false,
            "error": "Please select an option.",
            "id": "subGroup1Question2",
            "type": "choice",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice1",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice1",
                "value": "subGroup1Question3Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice2",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice2",
                "value": "subGroup1Question3Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice3",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice3",
                "value": "subGroup1Question3Choice3",
              },
            ],
            "currentAnswer": Array [],
            "custom": Object {
              "title": "subGroup1Question3",
            },
            "disabled": true,
            "error": "Please select at least 1 option.",
            "id": "subGroup1Question3",
            "type": "choices",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice1",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice1",
                "value": "subGroup1Question4Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice2",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice2",
                "value": "subGroup1Question4Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice3",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice3",
                "value": "subGroup1Question4Choice3",
              },
            ],
            "currentAnswer": Array [],
            "custom": Object {
              "title": "subGroup1Question3",
            },
            "disabled": false,
            "error": "Please select at least 1 option.",
            "id": "subGroup1Question4",
            "type": "choices",
            "validatedAnswer": undefined,
            "validating": false,
          },
        ],
      },
    ],
    "id": "group1",
    "questions": Array [
      Object {
        "choices": Array [],
        "currentAnswer": undefined,
        "custom": Object {
          "title": "group1Questions",
        },
        "disabled": false,
        "error": "This field is required.",
        "id": "group1Question1",
        "type": "any",
        "validatedAnswer": undefined,
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "custom": Object {
              "title": "group1Question2Choice1",
            },
            "disabled": false,
            "id": "group1Question2Choice1",
            "value": "group1Question2Choice1",
          },
          Object {
            "custom": Object {
              "title": "group1Question2Choice2",
            },
            "disabled": false,
            "id": "group1Question2Choice2",
            "value": "group1Question2Choice2",
          },
          Object {
            "custom": Object {
              "title": "group1Question2Choice3",
            },
            "disabled": false,
            "id": "group1Question2Choice3",
            "value": "group1Question2Choice3",
          },
        ],
        "currentAnswer": undefined,
        "custom": Object {
          "title": "group1Question2",
        },
        "disabled": false,
        "error": "Please select an option.",
        "id": "group1Question2",
        "type": "choice",
        "validatedAnswer": undefined,
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "custom": Object {
              "title": "group1Question3Choice1",
            },
            "disabled": false,
            "id": "group1Question3Choice1",
            "value": "group1Question3Choice1",
          },
          Object {
            "custom": Object {
              "title": "group1Question3Choice2",
            },
            "disabled": false,
            "id": "group1Question3Choice2",
            "value": "group1Question3Choice2",
          },
          Object {
            "custom": Object {
              "title": "group1Question3Choice3",
            },
            "disabled": false,
            "id": "group1Question3Choice3",
            "value": "group1Question3Choice3",
          },
        ],
        "currentAnswer": Array [],
        "custom": Object {
          "title": "group1Question3",
        },
        "disabled": false,
        "error": undefined,
        "id": "group1Question3",
        "type": "choices",
        "validatedAnswer": Array [],
        "validating": false,
      },
    ],
  },
]
`;

exports[`Invalid Answering setChoices empty answer 2`] = `
Object {
  "group1Question1": "This field is required.",
  "group1Question2": "Please select an option.",
  "subGroup1Question1": "This field is required.",
  "subGroup1Question2": "Please select an option.",
  "subGroup1Question4": "Please select at least 1 option.",
}
`;

exports[`Invalid Answering setChoices empty answer 3`] = `
Object {
  "group1Question3": Array [],
  "subGroup1Question3": Array [],
  "subGroup1Question4": Array [],
}
`;

exports[`Invalid Answering setChoices empty answer 4`] = `
Object {
  "group1Question3": Array [],
}
`;

exports[`Invalid Answering setChoices invalid choice 1`] = `
Array [
  Object {
    "custom": undefined,
    "disabled": false,
    "groups": Array [
      Object {
        "custom": undefined,
        "disabled": false,
        "groups": Array [],
        "id": "subGroup1",
        "questions": Array [
          Object {
            "choices": Array [],
            "currentAnswer": undefined,
            "custom": Object {
              "title": "subGroup1Questions",
            },
            "disabled": false,
            "error": "This field is required.",
            "id": "subGroup1Question1",
            "type": "any",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice1",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice1",
                "value": "subGroup1Question2Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice2",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice2",
                "value": "subGroup1Question2Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question2Choice3",
                },
                "disabled": false,
                "id": "subGroup1Question2Choice3",
                "value": "subGroup1Question2Choice3",
              },
            ],
            "currentAnswer": undefined,
            "custom": Object {
              "title": "subGroup1Question2",
            },
            "disabled": false,
            "error": "Please select an option.",
            "id": "subGroup1Question2",
            "type": "choice",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice1",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice1",
                "value": "subGroup1Question3Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice2",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice2",
                "value": "subGroup1Question3Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question3Choice3",
                },
                "disabled": true,
                "id": "subGroup1Question3Choice3",
                "value": "subGroup1Question3Choice3",
              },
            ],
            "currentAnswer": Array [],
            "custom": Object {
              "title": "subGroup1Question3",
            },
            "disabled": true,
            "error": "Please select at least 1 option.",
            "id": "subGroup1Question3",
            "type": "choices",
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice1",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice1",
                "value": "subGroup1Question4Choice1",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice2",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice2",
                "value": "subGroup1Question4Choice2",
              },
              Object {
                "custom": Object {
                  "title": "subGroup1Question4Choice3",
                },
                "disabled": false,
                "id": "subGroup1Question4Choice3",
                "value": "subGroup1Question4Choice3",
              },
            ],
            "currentAnswer": Array [],
            "custom": Object {
              "title": "subGroup1Question3",
            },
            "disabled": false,
            "error": "Please select at least 1 option.",
            "id": "subGroup1Question4",
            "type": "choices",
            "validatedAnswer": undefined,
            "validating": false,
          },
        ],
      },
    ],
    "id": "group1",
    "questions": Array [
      Object {
        "choices": Array [],
        "currentAnswer": undefined,
        "custom": Object {
          "title": "group1Questions",
        },
        "disabled": false,
        "error": "This field is required.",
        "id": "group1Question1",
        "type": "any",
        "validatedAnswer": undefined,
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "custom": Object {
              "title": "group1Question2Choice1",
            },
            "disabled": false,
            "id": "group1Question2Choice1",
            "value": "group1Question2Choice1",
          },
          Object {
            "custom": Object {
              "title": "group1Question2Choice2",
            },
            "disabled": false,
            "id": "group1Question2Choice2",
            "value": "group1Question2Choice2",
          },
          Object {
            "custom": Object {
              "title": "group1Question2Choice3",
            },
            "disabled": false,
            "id": "group1Question2Choice3",
            "value": "group1Question2Choice3",
          },
        ],
        "currentAnswer": undefined,
        "custom": Object {
          "title": "group1Question2",
        },
        "disabled": false,
        "error": "Please select an option.",
        "id": "group1Question2",
        "type": "choice",
        "validatedAnswer": undefined,
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "custom": Object {
              "title": "group1Question3Choice1",
            },
            "disabled": false,
            "id": "group1Question3Choice1",
            "value": "group1Question3Choice1",
          },
          Object {
            "custom": Object {
              "title": "group1Question3Choice2",
            },
            "disabled": false,
            "id": "group1Question3Choice2",
            "value": "group1Question3Choice2",
          },
          Object {
            "custom": Object {
              "title": "group1Question3Choice3",
            },
            "disabled": false,
            "id": "group1Question3Choice3",
            "value": "group1Question3Choice3",
          },
        ],
        "currentAnswer": Array [],
        "custom": Object {
          "title": "group1Question3",
        },
        "disabled": false,
        "error": undefined,
        "id": "group1Question3",
        "type": "choices",
        "validatedAnswer": Array [],
        "validating": false,
      },
    ],
  },
]
`;

exports[`Invalid Answering setChoices invalid choice 2`] = `
Object {
  "group1Question1": "This field is required.",
  "group1Question2": "Please select an option.",
  "subGroup1Question1": "This field is required.",
  "subGroup1Question2": "Please select an option.",
  "subGroup1Question4": "Please select at least 1 option.",
}
`;

exports[`Invalid Answering setChoices invalid choice 3`] = `
Object {
  "group1Question3": Array [],
  "subGroup1Question3": Array [],
  "subGroup1Question4": Array [],
}
`;

exports[`Invalid Answering setChoices invalid choice 4`] = `
Object {
  "group1Question3": Array [],
}
`;
